File: 22_专题/枚举/折半枚举/subsetSum/subsetSumSorted.ts
/**
 * O(2^n) 返回nums的各个子集的元素和的排序后的结果.
 */
function subsetSumSorted(arr: ArrayLike<number>): number[] {
  const n = arr.length
  let dp = [0]
  for (let i = 0; i < n; i++) {
    const ndp = dp.slice()
    for (let j = 0; j < dp.length; j++) {
      ndp[j] += arr[i]
    }
    dp = merge(dp, ndp)
  }
  return dp

  function merge(arr1: ArrayLike<number>, nums2: ArrayLike<number>): number[] {
    const n1 = arr1.length
    const n2 = nums2.length
    const res = Array<number>(n1 + n2)
    let i = 0
    let j = 0
    let k = 0

    while (i < n1 && j < n2) {
      if (arr1[i] < nums2[j]) res[k++] = arr1[i++]
      else res[k++] = nums2[j++]
    }

    while (i < n1) res[k++] = arr1[i++]
    while (j < n2) res[k++] = nums2[j++]
    return res
  }
}

export { subsetSumSorted }

File: 22_专题/枚举/折半枚举/subsetSum/SubsetSumSorted.go
// 返回nums的各个子集的元素和的排序后的结果
// !比求出所有的子集的元素和再排序要快很多

// func
package main

import "fmt"

func main() {
	fmt.Println(SubsetSumSorted([]int{1, 2}))
}

// 返回nums的各个子集的元素和的排序后的结果.
//  O(2^n)
func SubsetSumSorted(nums []int) []int {
	dp := []int{0}
	for _, v := range nums {
		ndp := make([]int, len(dp))
		for i, w := range dp {
			ndp[i] = w + v
		}
		dp = merge(dp, ndp)
	}
	return dp
}

func merge(a, b []int) []int {
	i, n := 0, len(a)
	j, m := 0, len(b)
	res := make([]int, 0, n+m)
	for {
		if i == n {
			return append(res, b[j:]...)
		}
		if j == m {
			return append(res, a[i:]...)
		}
		if a[i] < b[j] { // 改成 > 为降序
			res = append(res, a[i])
			i++
		} else {
			res = append(res, b[j])
			j++
		}
	}
}
